### ISSUE TITLE
Non causal sliding window mask ?

### ISSUE THREAD

**Issue by Optimox** [google/gemma_pytorch#73]
Hi,

I am trying to port gemma 2 to torchtune library.

When following the code for sliding window mask generation I first used a binary causal mask but it seems that this creates a non causal sliding window mask with 1s for previous tokens, 0s for future sliding window tokens and -inf for the other tokens.

I have shared a minimal reproducible code to demonstrate my problem [here](https://github.com/pytorch/torchtune/pull/1835#discussion_r1824896989).

Could someone clarify what format is expected for the causal or block causal masks in your implementation ?

I would also be curious to know why your are using `-2.3819763e38` instead of `-torch.inf` ?

Thank you for helping me clarifying the situation!

**Comment by Gopi-Uppari** [comment]
Hi @Optimox,

In the Gemma model's implementation, causal masks are utilized to ensure that each token in a sequence can only attend to itself and preceding tokens.
Expected Format for Causal Mask should Elements on or below the diagonal are set to 0 (indicating allowed attention), while elements above the diagonal 
are set to a large negative value -2.3819763e38 to effectively mask future tokens by assigning them negligible attention scores.

Use of -2.3819763e38 Instead of -torch.inf because numerical stability (Using extremely large negative values can prevent potential issues with floating-point precision during computations)
and compatibility reasons (Some hardware accelerators or libraries may not handle -inf gracefully, leading to undefined behavior or errors.)

For more reference, could you please refer to this [reference](https://github.com/google/gemma_pytorch/blob/main/gemma/model.py)

Thank you.

**Comment by Gopi-Uppari** [comment]
Hi @Optimox,

Could you please confirm if this issue is resolved for you with the above comment ? Please feel free to close the issue if it is resolved ?

Thank you.

**Comment by Optimox** [comment]
Yes thank you for your help!

**Comment by ebsmothers** [comment]
Hi @Gopi-Uppari sorry I am still a bit confused on this. I am less concerned about the usage of `-2.3819763e38` vs `-torch.inf` but more confused about why there would be unmasked outputs above the diagonal. For instance if I directly copy-paste [this snippet](https://github.com/google/gemma_pytorch/blob/main/gemma/model.py#L312-L316) and inspect the outputs I see the following mask:

```
import torch

s_x = 5
sliding_window_size = 3
mask = torch.tril(
                torch.ones(
                    size=(s_x, s_x),
                    dtype=torch.bool,
                )
            )

all_ones = torch.ones_like(mask)
sliding_mask = torch.triu(
    all_ones, -1 * sliding_window_size + 1
) * torch.tril(all_ones, sliding_window_size - 1)
mask = torch.where(sliding_mask == 1, mask, -2.3819763e38)
print(mask)
...

tensor([[ 1.0000e+00,  0.0000e+00,  0.0000e+00, -2.3820e+38, -2.3820e+38],
        [ 1.0000e+00,  1.0000e+00,  0.0000e+00,  0.0000e+00, -2.3820e+38],
        [ 1.0000e+00,  1.0000e+00,  1.0000e+00,  0.0000e+00,  0.0000e+00],
        [-2.3820e+38,  1.0000e+00,  1.0000e+00,  1.0000e+00,  0.0000e+00],
        [-2.3820e+38, -2.3820e+38,  1.0000e+00,  1.0000e+00,  1.0000e+00]])
```
